### ISSUE TITLE
Alignment loss function

### ISSUE THREAD

**Issue by marcpaga** [google/deepconsensus#30]
Thanks for the very interesting work. 

I was wondering about the alignment loss used to train the model. It is clear that indels can shift the whole predicted sequence and then a loss like cross-entropy explodes by small mistakes. I thought a CTC loss would work in this scenario, but you developed a new alignment loss for this task. I was wondering if you could elaborate on why this alignment loss is needed or why CTC is not viable here.

**Comment by kishwarshafin** [comment]
Hi @marcpaga ,

Thanks for opening this issue. There are several things to consider here:

* CTC-loss takes all possible alignments and then sums the alignment scores based on the predictions. In calculating distance between two genome sequences that are very similar, this is not ideal as suboptimal alignments can contribute heavily to the prediction. You can read in our [manuscript](https://www.biorxiv.org/content/10.1101/2021.08.31.458403v1.full.pdf) that we allow special handling for insertions and deletions (gap penalty) and maximize over the best alignment score. This way the loss is smoother as we control how much the suboptimal alignments can contribute to the loss using a parameter.

* CTC-loss does and exhaustive search over all possible alignments and is a bit slower. The alignment-loss function implemented for deepconsensus provides an efficient implementation for faster training on GPUs and TPUs.

Finally, we have not exhaustively looked at all the loss functions that are available. So, future experiments will other alignment-based loss functions including CTC may give us observations to help answer your question more accurately.
